packages/graph/hash_graph/lib/graph/src/shared/identifier/mod.rs

```diff
@@ -13,14 +23,177 @@ pub mod account;
+    end: Bound<DateTime<Utc>>,
+}
+
+impl FromSql<'_> for Timespan {
+    fn from_sql(_: &Type, buf: &[u8]) -> Result<Self, Box<dyn std::error::Error + Sync + Send>> {
+        fn parse_bound(
+            bound: &RangeBound<Option<&[u8]>>,
+        ) -> Result<Bound<DateTime<Utc>>, Box<dyn std::error::Error + Send + Sync>> {
+            Ok(match bound {
+                RangeBound::Inclusive(None) | RangeBound::Exclusive(None) => {
+                    unimplemented!("null ranges are not supported")
+                }
+                RangeBound::Inclusive(Some(bytes)) => {
+                    let timestamp = DateTime::from_sql(&Type::TIMESTAMPTZ, bytes)?;
+                    Bound::Included(timestamp)
+                }
+                RangeBound::Exclusive(Some(bytes)) => {
+                    let timestamp = DateTime::from_sql(&Type::TIMESTAMPTZ, bytes)?;
+                    Bound::Excluded(timestamp)
+                }
+                RangeBound::Unbounded => Bound::Unbounded,
+            })
+        }
+
+        fn is_negative_infinity(bound: &RangeBound<Option<&[u8]>>) -> io::Result<bool> {
+            Ok(match bound {
+                RangeBound::Inclusive(Some(mut bytes)) | RangeBound::Exclusive(Some(mut bytes)) => {
+                    bytes.read_i64::<BigEndian>()? == i64::MIN
+                }
+                _ => false,
+            })
+        }
+
+        fn is_positive_infinity(bound: &RangeBound<Option<&[u8]>>) -> io::Result<bool> {
+            Ok(match bound {
+                RangeBound::Inclusive(Some(mut bytes)) | RangeBound::Exclusive(Some(mut bytes)) => {
+                    bytes.read_i64::<BigEndian>()? == i64::MAX
+                }
+                _ => false,
+            })
+        }
+
+        match postgres_protocol::types::range_from_sql(buf)? {
+            Range::Empty => unimplemented!("Empty ranges are not supported"),
+            Range::Nonempty(lower, upper) => Ok(Self {
+                start: if is_negative_infinity(&lower)? {
+                    Bound::Unbounded
+                } else {
+                    parse_bound(&lower)?
+                },
+                end: if is_positive_infinity(&upper)? {
+                    Bound::Unbounded
+                } else {
+                    parse_bound(&upper)?
+                },
+            }),
+        }
+    }
+
+    fn accepts(ty: &Type) -> bool {
+        matches!(ty, &Type::TSTZ_RANGE)
+    }
+}
+
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, Serialize, Deserialize)]
+pub struct DecisionTimespan {
+    start_bound: Bound<DecisionTimestamp>,
+    end_bound: Bound<DecisionTimestamp>,
+}
+
+impl DecisionTimespan {
+    #[must_use]
+    pub(crate) fn new(timespan: Timespan) -> Self {
+        Self {
+            start_bound: timespan.start.map(DecisionTimestamp),
+            end_bound: timespan.end.map(DecisionTimestamp),
+        }
+    }
+}
+
+impl RangeBounds<DecisionTimestamp> for DecisionTimespan {
```

Comment:
How convenient that the `Bound` would be the implementation detail used for BTree ranges which happen to be the kind of index we're using for the spans (not that it would make any difference) :D